src/plugins/RelightHelpers.js

```diff
@@ -0,0 +1,228 @@
+/**
+ * Generates an object containing a full tile pyramid. Mirador only generates the parts that are needed
+ * as they are loaded in, so you cannot access the full pyramid until all the tiles have loaded.  We need the
+ * full data structure earlier so here we calculate it.  The algorithm is based on Tom Crane's tile-exploder
+ * @param {string} id IIIF image resource id
+ * @param {number} width width of the image resource at a particular scale factor
+ * @param {number} height height of the image resource at a particular scale factor
+ * @param {object} tileSource object containing information parsed out from the image info.json in the manifest
+ * @param {number} scaleFactor the scaleFactor you want to generate tiles for
+ * @returns {{}} an array of objects containing IIIF image tile URLs with tile dimensions
+ */
+export function generateTiles(id, width, height, tileSource, scaleFactor) {
+  let tiles = [];
+  const tileWidth = tileSource.width;
+  const tileHeight = tileSource.height || tileSource.width;
+  const factors = tileSource.scaleFactors.sort(function (a, b) {
+    return b - a;
+  }); // spec doesn't actually specify order
+
+  let scale = factors[scaleFactor];
+  const regionWidth = scale * tileWidth;
+  const regionHeight = scale * tileHeight;
+  const scaleWidth = Math.ceil(width / scale);
+  const scaleHeight = Math.ceil(height / scale);
+  let y = 0;
+
+  while (y < height) {
+    let x = 0;
+    while (x < width) {
+      let region;
+      let rw;
+      let rh;
+
+      if (scaleWidth <= tileWidth && scaleHeight <= tileHeight) {
+        rw = tileWidth;
+        rh = tileHeight;
+        region = 'full';
+      } else {
+        rw = Math.min(regionWidth, width - x);
+        rh = Math.min(regionHeight, height - y);
+        region = x + ',' + y + ',' + rw + ',' + rh;
+      }
+      const scaledWidthRemaining = Math.ceil((width - x) / scale);
+      const tw = Math.min(tileWidth, scaledWidthRemaining);
+      const iiifArgs = '/' + region + '/' + tw + ',/0/default.jpg';
```

Comment:
I think we might need to track rotation here. you can rotate the images using mirador image tools, and then enabling the plugin gives you this: 
![image](https://github.com/bodleian/archiox-mirador-plugin/assets/390869/3650aeeb-720b-4b8c-8ce6-2373162352a2)

Reply:
Actually come to think of it, here is not the place to track it. so ignore me and i'll add a seperate issue!